src/Services/InventoryStatusService.php

```diff
@@ -0,0 +1,408 @@
+  const LAST_COMPLETION = 'lastCompletion';
+  const LAST_ATTEMPT_TIMESTAMP = 'lastAttemptTimestamp';
+  const LAST_ATTEMPT_SUCCEEDED = 'lastAttemptSucceeded';
+
+  const STATE_RUNNING = 'running';
+  const STATE_IDLE = 'idle';
+
+  const INVENTORY_CRON_STATUS_FULL = 'full_inventory_cron_status';
+  const INVENTORY_STATUS_UPDATED_AT_FULL = 'full_inventory_status_updated_at';
+  const INVENTORY_LAST_COMPLETION_FULL = "full_inventory_last_completion";
+  const INVENTORY_LAST_ATTEMPT_FULL = "full_inventory_last_attempt";
+  const INVENTORY_SUCCESS_FULL = "full_inventory_success";
+  const INVENTORY_LAST_SUCCESS_START_FULL = "full_inventory_last_success_start";
+
+  const KEYS_FOR_FULL = [
+    self::INVENTORY_CRON_STATUS_FULL,
+    self::INVENTORY_STATUS_UPDATED_AT_FULL,
+    self::INVENTORY_LAST_COMPLETION_FULL,
+    self::INVENTORY_SUCCESS_FULL,
+    self::INVENTORY_LAST_SUCCESS_START_FULL
+  ];
+
+  const INVENTORY_CRON_STATUS_PARTIAL = 'partial_inventory_cron_status';
+  const INVENTORY_STATUS_UPDATED_AT_PARTIAL = 'partial_inventory_status_updated_at';
+  const INVENTORY_LAST_COMPLETION_PARTIAL = "partial_inventory_last_completion";
+  const INVENTORY_LAST_ATTEMPT_PARTIAL = "partial_inventory_last_attempt";
+  const INVENTORY_SUCCESS_PARTIAL = "partial_inventory_success";
+  const INVENTORY_LAST_SUCCESS_START_PARTIAL = "partial_inventory_last_success_start";
+
+  const KEYS_FOR_PARTIAL = [
+    self::INVENTORY_CRON_STATUS_PARTIAL,
+    self::INVENTORY_STATUS_UPDATED_AT_PARTIAL,
+    self::INVENTORY_LAST_COMPLETION_PARTIAL,
+    self::INVENTORY_LAST_ATTEMPT_PARTIAL,
+    self::INVENTORY_SUCCESS_PARTIAL,
+    self::INVENTORY_LAST_SUCCESS_START_PARTIAL
+  ];
+
+  /**
+   * @var KeyValueRepository
+   */
+  private $keyValueRepository;
+
+  /**
+   * @var LoggerContract
+   */
+  private $logger;
+
+  /**
+   * InventoryStatusService constructor.
+   *
+   * @param KeyValueRepository $keyValueRepository
+   * @param LoggerContract $logger
+   */
+  public function __construct(
+    KeyValueRepository $keyValueRepository,
+    LoggerContract $logger
+  ) {
+    $this->keyValueRepository = $keyValueRepository;
+    $this->logger = $logger;
+  }
+
+  /**
+   * Set the global status for full or partial inventory syncing,
+
+   * returning the old state.
+   * @param bool $full
+   * @param string $statusValue
+   * @param string $timestamp
+   * @return string
+   */
+  private function setServiceStatusValue($full, $statusValue, $timestamp = null): string
+  {
+    if (!isset($timestamp) || empty($timestamp)) {
+      $timestamp = self::getCurrentTimestamp();
+    }
+
+    $statusKey = $full ? self::INVENTORY_CRON_STATUS_FULL : self::INVENTORY_CRON_STATUS_PARTIAL;
+    $logKeyStateChange = $full ? self::LOG_KEY_STATE_CHANGE_FULL : self::LOG_KEY_STATE_CHANGE_PARTIAL;
+
+    $oldStatus = $this->getServiceStatusValue($full);
+
+    if ($oldStatus != $statusValue) {
+      $this->keyValueRepository->putOrReplace($statusKey, $statusValue);

```

Comment:
What happen when full inventory sync and partial inventory sync run at the same time? Looks like right now we are allowing that.

Reply:
Check out the top of the do/while loop. I am attempting to gracefully quit anything "old" that is running:

```
  $mostRecentFullStart = $this->statusService->getLastAttemptTime(true);

        if (
          isset($mostRecentFullStart) && !empty($mostRecentFullStart)
          && strtotime($mostRecentFullStart) > strtotime($startTimeStamp)
        ) {
          // a sync of all items is happening AFTER this one, so this sync is stale!
          throw new InventorySyncInterruptedException("Inventory sync started at " . $startTimeStamp .
            " lost priority to Full Inventory sync started at " . $mostRecentFullStart);
        }

        $mostRecentPartialStart = $this->statusService->getLastAttemptTime(false);

        if (
          !$fullInventory
          && isset($mostRecentPartialStart) && !empty($mostRecentPartialStart)
          && strtotime($mostRecentPartialStart) > strtotime($startTimeStamp)
        ) {
          // this is a partial inventory, but another partial inventory started after it.
          // this one should stop running.
          throw new InventorySyncInterruptedException("Inventory sync started at " . $startTimeStamp .
            " lost priority to Inventory sync started at " . $mostRecentPartialStart);
        }
```

Reply:
If we just allow only one of sync service(full inventory sync and partial inventory sync) run at a time
we can just use one statestatus for both full inventory sync and partial inventory sync
then we don't need to do the above check as well

Reply:
yeah I was thinking about that but it conflicted with the fact that we want the UI to allow the user to immediately start a Full sync without having to wait for any current partial sync to end. totally open to ideas.

Reply:
I have concern about the partial sync running and we kick off full sync. it  may cause some unexpected behavior, e.g. race condition.
Is there a hard requirement that we have to kick off full inventory immediately? Or we can just pop user some warning messages to tell user that there is an inventory sync ongoing and ask user to wait for a few seconds to retry?

Reply:
That is going to add some level of difficulty to the UI. I'll think about it though.
We already tell them when a Full sync is currently running because that's a valid reason they should wait or abort their attempt to do a full sync. Partial sync is not a valid reason to prevent the user from doing anything.

Reply:
I wonder if we can reduce the risk of a full sync starting while an ongoing partial sync is happening by having the full sync still update it's status as running and then wait until the partial sync detects this and exits.

Reply:
Considering the strong feelings I'm seeing here, I'm tempted to change the `Full Inventory` tab in the UI to be `Inventory`, then do as @jiafanwu is saying and track Inventory as a singular thing. 

- I would change the `running` status to be `full` or `partial`
- The UI would say `syncing all inventory` or `syncing changes since 08-01-2020 13:50 +2:00`
- There would be a new `sync recent changes` button, and the user would pick between that and the `sync all inventory` option
- All `sync` buttons would be blocked whilst the status is not `idle.`

Reply:
I like that approach. Could we first make the backend changes and update the `Full Inventory` tab to just block whenever a full or partial is detected? Then we could get this fix out quicker while you work on the full UI changes?

Reply:
eh the backend changes in this PR are in a bit of a convoluted state, which is probably what concerned @jiafanwu. I'm thinking I'll be in "beta" on the new-new backend and new UI sometime tomorrow, and if not, I can clean up this PR to do as you're suggesting.